compiler/luci/service/src/Nodes/CircleReshape.cpp

```diff
@@ -65,93 +41,121 @@ luci::CircleNode *CloneNodeLet<CN::OPQR>::visit(const luci::CircleReshape *node)
 namespace sinf
 {
 
+/**
+ * @note  CircleReshape always has two inputs: `tensor` and `shape`.
+ *        The `shape` input can be CircleConst, CircleOutputDummy, or CircleNode.
+ *        - If the `shape` input is CircleConst, the shape is inferred from the constant.
+ *        - If the `shape` input is CircleOutputDummy, the shape is inferred from
+ *          the attribute if it exists. If the attribute does not exist,
+ *          the shape is inferred from the node iteself.
+ *        - If the `shape` input is CircleNode, the dynamic shape is propagated.
+ */
 loco::TensorShape Algorithm::visit(const luci::CircleReshape *node)
 {
-  LOGGER(l);
-
   const loco::DataType S32 = loco::DataType::S32;
 
-  loco::TensorShape shape_by_input;
-  {
-    LUCI_ASSERT(node->shape(), "2nd input shape() should not be nullptr");
+  // CircleReshape node must have `shape` input
+  LUCI_ASSERT(node->shape(), "2nd input shape() should not be nullptr");
 
-    // Only support node's shape() is CircleConst with S32
-    // TODO support other node with other types
-    auto const_shape_node = dynamic_cast<luci::CircleConst *>(node->shape());
-    if (const_shape_node != nullptr)
+  bool should_infer = true;
+  loco::TensorShape output_shape;
+  {
```

Comment:
Below if-else statements can be little bit more simplified with following codes :)

```cpp
if (auto const_shape = dynamic_cast<luci::CircleConst *>(node->shape()))
{
  // inference with const_shape
}
else if (auto dummy_shape = dynamic_cast<luci::CircleOutputDummy *>(node->shape()))
{
  // inference with attribute
}
else
{
  auto node_shape = loco::must_cast<luci::CircleNode *>(node->shape())
  // propagate dynamic shape
}
```